Exponentially back off the time between message polling instead of do…

picklesrus opened the pull request:
…ing it every two minutes.

### Resolves:

https://github.com/LLK/scratch-www/issues/4079

### Changes:

Back off how often we poll for new messages

### Test Coverage:

There are unit tests for the new setupMessagePolling function.

benjiwheeler commented on src/components/navigation/www/navigation.jsx:
```diff
@@ -77,6 +69,23 @@ class Navigation extends React.Component {
         return `/users/${this.props.user.username}/`;
     }
+
+    setupMessagePolling (minutes) {
```
if it's not too much trouble, it might make more sense to use seconds as the unit -- that seems to be more of a convention than minutes.

picklesrus replied:
Good point! This name is left over from a previous iteration. Renamed to pollForMessages since that's actually what it is doing.

benjiwheeler commented on src/components/navigation/www/navigation.jsx:
```diff
@@ -77,6 +69,23 @@ class Navigation extends React.Component {
+            const nextFetch = minutes * 2;
+            const timeoutId = setTimeout(() => {
+                this.setupMessagePolling(nextFetch);
+            }, nextFetch * 60000);
```
The convention you used earlier, on line 47, where you showed the multiplication that got you this number ("2 * 60 * 1000") might also be good here

picklesrus replied:
I changed everything over to being in ms as you suggested. Since I did that, nextFetch is now in millis so I don't need the multiplication anymore.

benjiwheeler commented on src/components/navigation/www/navigation.jsx:
```diff
@@ -27,35 +27,27 @@ class Navigation extends React.Component {
+            'setupMessagePolling'
         ]);
         this.state = {
             messageCountIntervalId: -1 // javascript method interval id for getting messsage count.
```
this is maybe going beyond the scope of this change, but I don't think that messageCountIntervalId belongs inside state. There's no need for the component to re-render just because messageCountIntervalId changed. 

There are a few examples of how you might just store the response from `setTimeout()` in a regular member variable: in www, src/components/modal/social/container.jsx does this, and in gui, src/components/action-menu/action-menu.jsx does this.

But I totally understand if you don't want to make that change as part of this change!

picklesrus replied:
I was swamped this week so didn't have time to look into this. If you don't have time to review before you head out of office, maybe I'll take a look next week. Otherwise, I suggest wee keep it for a follow up.

benjiwheeler commented:
this looks good, just a few suggestions! Way to go getting those tests working

picklesrus commented:
I moved the intervalId tracking out of the state object to a member variable. Also added some tests -- glad I did because they caught a bug that would have been hard to find otherwise :)

benjiwheeler commented on src/components/navigation/www/navigation.jsx:
```diff
@@ -27,56 +27,60 @@ class Navigation extends React.Component {
+        // We only poll if it has been less than 30 minutes.
+        // Chances of someone actively using the page for that long without
+        // a navigation is low.
+        if (ms < 32 * 60 * 1000) { // 32 minutes
```
This could be 30 instead of 32 to match the comment, but it's not a big deal

benjiwheeler commented on src/components/navigation/www/navigation.jsx:
```diff
@@ -27,56 +27,60 @@ class Navigation extends React.Component {
             } else {
                 // clear message count check, and set to default id.
-                clearInterval(this.state.messageCountIntervalId);
+                clearTimeout(this.messageCountTimeoutId);
```
I think ideally we'd test messageCountTimeoutId to see if it's !== -1 like on line 59. But it's not a big deal because I'm pretty sure JS will just ignore it if we call clearTimeout(-1) or call it with a stale function id

benjiwheeler commented:
LGTM! Two small notes, but this can go in without my needing to review again!
